addons/sale_coupon/models/coupon_program.py

```diff
@@ -283,23 +152,29 @@ def _filter_programs_from_common_rules(self, order, next_order=False):
             programs += programs_curr_order._filter_not_ordered_reward_programs(order)
         return programs
 
-    def _is_valid_partner(self, partner):
-        if self.rule_partners_domain:
-            domain = safe_eval(self.rule_partners_domain) + [('id', '=', partner.id)]
-            return bool(self.env['res.partner'].search_count(domain))
-        else:
-            return True
+    def _is_global_discount_program(self):
+        self.ensure_one()
+        return self.promo_applicability == 'on_current_order' and \
+               self.reward_type == 'discount' and \
+               self.discount_type == 'percentage' and \
+               self.discount_apply_on == 'on_order'
 
-    def _is_valid_product(self, product):
-        # NOTE: if you override this method, think of also overriding _get_valid_products
-        if self.rule_products_domain:
-            domain = safe_eval(self.rule_products_domain) + [('id', '=', product.id)]
-            return bool(self.env['product.product'].search_count(domain))
-        else:
-            return True
+    def _keep_only_most_interesting_auto_applied_global_discount_program(self):
+        '''Given a record set of programs, remove the less interesting auto
+        applied global discount to keep only the most interesting one.
+        We should not take promo code programs into account as a 10% auto
+        applied is considered better than a 50% promo code, as the user might
+        not know about the promo code.
+        '''
+        programs = self.filtered(lambda p: p._is_global_discount_program() and p.promo_code_usage == 'no_code_needed')
+        if not programs: return self
+        most_interesting_program = max(programs, key=lambda p: p.discount_percentage)
+        # remove least interesting programs
+        return self - (programs - most_interesting_program)
 
-    def _get_valid_products(self, products):
-        if self.rule_products_domain:
-            domain = safe_eval(self.rule_products_domain) + [('id', 'in', products.ids)]
-            return self.env['product.product'].search(domain)
-        return products
+    def _discount_line_product_vals(self, name):
```

Comment:
If we want to avoid that, we could make `coupon` depend on `account`. I have no issue with doing that, since both PoS and Sales already do...

If we don't do that, you'll need to put this override also in `pos_coupon`...

Having twice the same override for the same 'problem' because we didn't put `account` as a dep of a module that doesn't do anything by itself seems like a waste imho